process/block/alteredaccounts/alteredAccountsProvider.go

```diff
@@ -0,0 +1,226 @@
+	alteredAccounts map[string]*indexer.AlteredAccount,
+) error {
+	addressBytes := []byte(addressStr)
+	encodedAddress := aap.addressConverter.Encode(addressBytes)
+
+	account, err := aap.accountsDB.LoadAccount(addressBytes)
+	if err != nil {
+		log.Warn("cannot load account when computing altered accounts",
+			"address", encodedAddress,
+			"error", err)
+		return err
+	}
+
+	userAccount, ok := account.(state.UserAccountHandler)
+	if !ok {
+		log.Warn("cannot cast AccountHandler to UserAccountHandler", "address", encodedAddress)
+		return err
+	}
+
+	alteredAccounts[encodedAddress] = &indexer.AlteredAccount{
+		Address: encodedAddress,
+		Balance: userAccount.GetBalance().String(),
+		Nonce:   userAccount.GetNonce(),
+	}
+
+	for tokenKey, tokenData := range markedAccountTokens {
+		err = aap.addTokensDataForMarkedAccount([]byte(tokenKey), encodedAddress, userAccount, tokenData, alteredAccounts)
+		if err != nil {
+			log.Warn("cannot fetch token data for marked account", "error", err)
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (aap *alteredAccountsProvider) addTokensDataForMarkedAccount(
+	tokenKey []byte,
+	encodedAddress string,
+	userAccount state.UserAccountHandler,
+	markedAccountToken *markedAlteredAccountToken,
+	alteredAccounts map[string]*indexer.AlteredAccount,
+) error {
+	nonce := markedAccountToken.nonce
+	tokenID := markedAccountToken.identifier
+
+	storageKey := []byte(core.ElrondProtectedKeyPrefix + core.ESDTKeyIdentifier)
+	storageKey = append(storageKey, tokenKey...)
+
+	esdtTokenBytes, err := userAccount.RetrieveValueFromDataTrieTracker(storageKey)
+	if err != nil {
+		return err
+	}
+
+	var esdtToken esdt.ESDigitalToken
+	err = aap.marshalizer.Unmarshal(&esdtToken, esdtTokenBytes)
+	if err != nil {
+		return err
+	}
+
+	alteredAccount := alteredAccounts[encodedAddress]
+
+	alteredAccount.Tokens = append(alteredAccount.Tokens, &indexer.AccountTokenData{
+		Identifier: tokenID,
+		Balance:    esdtToken.Value.String(),
+		Nonce:      nonce,
+		MetaData:   esdtToken.TokenMetaData,
+	})
+
+	alteredAccounts[encodedAddress] = alteredAccount
+
+	return nil
+}
+
+func (aap *alteredAccountsProvider) extractAddressesWithBalanceChange(
+	txPool *indexer.Pool,
+	markedAlteredAccounts map[string]*markedAlteredAccount,
+) {
+	selfShardID := aap.shardCoordinator.SelfId()
+
+	aap.extractAddressesFromTxsHandlers(selfShardID, txPool.Txs, markedAlteredAccounts)
+	aap.extractAddressesFromTxsHandlers(selfShardID, txPool.Scrs, markedAlteredAccounts)
+	aap.extractAddressesFromTxsHandlers(selfShardID, txPool.Rewards, markedAlteredAccounts)
+	aap.extractAddressesFromTxsHandlers(selfShardID, txPool.Invalid, markedAlteredAccounts)
+}
+
+func (aap *alteredAccountsProvider) extractAddressesFromTxsHandlers(
+	selfShardID uint32,
+	txsHandlers map[string]data.TransactionHandler,
+	markedAlteredAccounts map[string]*markedAlteredAccount,
+) {
+	for _, txHandler := range txsHandlers {
+		senderAddress := txHandler.GetSndAddr()
```

Comment:
I think you should ignore the sender of the rewards transactions. 
Also, you should ignore the receiver of the invalid transactions.

Reply:
i've added a check for invalid addresses. if the address is not valid, don't mark it as altered account